ports/stm32f4/common-hal/pulseio/PWMOut.c

```diff
@@ -0,0 +1,426 @@
+            //check if the timer has a channel active
+            if (reserved_tim[mcu_tim_pin_list[i].tim_index-1] != 0) {
+                //is it the same channel? (or all channels reserved by a var-freq)
+                if (reserved_tim[mcu_tim_pin_list[i].tim_index-1] & 1<<(mcu_tim_pin_list[i].channel_index-1)) {
+                    tim_chan_taken = true;
+                    continue; //keep looking, might be another viable option
+                }
+                //If the frequencies are the same it's ok
+                if (tim_frequencies[mcu_tim_pin_list[i].tim_index-1] != frequency) {
+                    tim_taken_f_mismatch = true;
+                    continue; //keep looking
+                }
+                //you can't put a variable frequency on a partially reserved timer
+                if (variable_frequency) {
+                    var_freq_mismatch = true;
+                    continue; //keep looking
+                }
+                first_time_setup = false; //skip setting up the timer
+            }
+            //No problems taken, so set it up
+            self->tim = &mcu_tim_pin_list[i];
+            break;
+        }
+    }
+
+    //handle valid/invalid timer instance
+    if (self->tim!=NULL) {
+        //create instance
+        TIMx = mcu_tim_banks[self->tim->tim_index-1];
+        //reserve timer/channel
+        if (variable_frequency) {
+            reserved_tim[self->tim->tim_index-1] = 0x0F;
+        } else {
+            reserved_tim[self->tim->tim_index-1] |= 1<<(self->tim->channel_index-1);
+        }
+        tim_frequencies[self->tim->tim_index-1] = frequency;
+    } else { //no match found
+        if (tim_chan_taken) {
+            mp_raise_ValueError(translate("No more timers available on this pin."));
+        } else if (tim_taken_f_mismatch) {
+            mp_raise_ValueError(translate("Frequency must be the same as as the existing PWMOut using this timer"));
+        } else if (var_freq_mismatch) {
+            mp_raise_ValueError(translate("Cannot vary frequency on a timer that is already in use"));
+        } else {
+            mp_raise_ValueError(translate("Invalid pins"));
+        }
+    }
+
+    GPIO_InitTypeDef GPIO_InitStruct = {0};
+    GPIO_InitStruct.Pin = pin_mask(pin->number);
+    GPIO_InitStruct.Mode = GPIO_MODE_AF_PP;
+    GPIO_InitStruct.Pull = GPIO_NOPULL;
+    GPIO_InitStruct.Speed = GPIO_SPEED_FREQ_HIGH;
+    GPIO_InitStruct.Alternate = self->tim->altfn_index;
+    HAL_GPIO_Init(pin_port(pin->port), &GPIO_InitStruct);
+
+    //TODO: factor all of these into periph.c?
+    tim_clock_enable(1<<(self->tim->tim_index - 1));
+
+    //translate channel into handle value
+    switch (self->tim->channel_index) {
+        case 1: self->channel = TIM_CHANNEL_1; break;
+        case 2: self->channel = TIM_CHANNEL_2; break;
+        case 3: self->channel = TIM_CHANNEL_3; break;
+        case 4: self->channel = TIM_CHANNEL_4; break;
+    }
```

Comment:
`self->channel = 1 << (2 * (self->tim->channel_index - 1))`

https://github.com/adafruit/stm32f4/blob/89a356f0b44883d55c45a5d22940777a44a2b881/STM32F4xx_HAL_Driver/Inc/stm32f4xx_hal_tim.h#L693

Reply:
has to be `self->channel = 4 * (self->tim->channel_index - 1)` I think. I thought it was weird that it wasn't a bit shift here, which is why I linked the macros. But on double checking it looks like it's consistent across devices so this should be ok.